Implement "Battlefield Camera" setting and smooth zooming

pchote opened the pull request:
This PR builds on #17308 and its prerequisites to finally implement the world "camera height" / custom resolution setting outlined in #10382. This also adds smooth mouse-wheel zooming and makes several changes to the zoom hotkeys and configuration to fit with the new and next changes.

A key part of this PR is an implementation of a custom scaling filter (inspired by [this article](https://colececil.io/blog/2017/scaling-pixel-art-without-destroying-it/)) that resamples the world frame buffer without excessive blurring or artifacting.

The focus for this PR is to support players (myself included) who have monitors where the native scaling is uncomfortably small, but pixel doubling is uncomfortably large. I've tried to take into account the feedback from the Discord discussion I started a few months ago, and think I've found a good solution that improves the default configuration for many players while giving people who prefer the old behaviour ways to preserve it.

There are several scope creeps that I considered but ruled out from implementing in this PR:

* Supporting zooms < 1, aside from keeping compatibility with the existing ability for spectators and the map editor - this requires deeper changes to the viewport scrolling code to work well, and opens up a can of wormy gameplay implications. More discussion and experimentation is required.
* Supporting zoom ranges more than just 1-2 - this requires deeper thought and community relations wrt the impact on the toggle zoom (now reset zoom) hotkey and breaking/changing muscle memory.
* Allowing missions / lobby options to restrict the maximum viewport size - this one should be fairly  safe, but is better as its own PR.
* Implementing the UI scaling half of #10382 - this is another large pile of work, which I will be continuing with in the next set of PRs.
* Scaling health bars etc - this is part of the UI scaling, above.


Depends on ~#17308~, ~#17431~, ~#17435.~

People on Windows with their DPI/desktop scaling not at 100% may need to disable `Graphics.DisableWindowsDPIScaling` to see the proper effect of this PR. This setting will be removed as part of the UI scaling overhaul.

pchote committed:
Implement optimized pixel-art antialiasing mode for non-integer world zoom.

pchote commented on OpenRA.Game/Settings.cs:
```diff
@@ -221,7 +222,6 @@ public class GameSettings
 		public string AuthProfile = "player.oraid";
 
 		public bool AllowZoom = true;
-		public Modifiers ZoomModifier = Modifiers.Ctrl;
```
I originally pushed for this modifier because the previous solution wasn't a "real" zoom, so shouldn't have owned the mousewheel by default. This is no longer true, so IMO we can benefit more by dropping the option. This PR is promoting zooming to a top-tier feature, and the changes to the zoom reset hotkey mean that scrolling is now the only way to access the additional zoom levels in spectator/editor modes.

tovl replied:
In that case I think we should remove the `AllowZoom` checkbox as well then.

Regarding accessing the additional zoomlevels without the mouse wheel: We could use the '+' and '-' keys for that.

abmyii replied:
As per our discussion on Discord, could this be optional?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -184,14 +184,20 @@ void RegisterSettingsPanel(PanelType type, Func<Widget, Action> init, Func<Widge
 
+		static readonly Dictionary<WorldViewport, string> ViewportSizeNames = new Dictionary<WorldViewport, string>()
+		{
+			{ WorldViewport.Close, "Close" },
```
Close is ~classic game size
Medium is ~720p
Far is ~1080p
Native/Furthest is always a zoom of 1.

The "~" is because the game will try to pick a size that is near but not necessarily matching these so we don't unnecessarily degrade the visual quality at different aspect ratios.

pchote commented on mods/common/chrome/editor.yaml:
```diff
@@ -584,23 +586,6 @@ Container@EDITOR_WORLD_ROOT:
-			Align: Right
-			Font: Bold
-			Contrast: true
-		DropDownButton@ZOOM_BUTTON:
```
Specific named zoom levels don't make a lot of sense now, so this is instead moved to the mouse wheel.

tovl replied:
This creates a gap in the button row, so the undo/redo buttons should be moved to the left.

pchote commented:
Rebased and fixed a shader compilation error on non-macOS.

tovl commented:
This looks really great in general.

Some issues I noticed while testing:
- Sometimes when spectating in TS at the closest setting I get some weird artifacts: Black jagged blocks at the bottom of the screen. It seems to be map dependent; I get it consistently on Tactical.
- I think the scroll zoom could benefit from having an acceleration curve.
- At the closest setting there is a noticeable wobble when zooming.

tovl commented on OpenRA.Game/Graphics/Viewport.cs:
```diff
@@ -140,8 +167,75 @@ public Viewport(WorldRenderer wr, Map map)
+				// TODO: Allow zooming out until the full map is visible
+				// We need to improve our viewport scroll handling to center the map as we zoom out
+				// before this will work well enough to enable
+				unlockedMinZoom = minZoom / unlockMinimumZoomScale;
```
Can we please invert the meaning of `unlockMinimumZoomScale` so it's meaning is consistent with `MaxZoomScale`?

tovl commented on OpenRA.Game/WorldViewportSizes.cs:
```diff
@@ -0,0 +1,33 @@
+
+		public readonly float MaxZoomScale = 2.0f;
+		public readonly int MaxZoomWindowHeight = 240;
+		public readonly bool AllowNativeZoom = true;
```
Can we please make these settings yaml configurable?

pchote replied:
They already are. `IGlobalModData` implementations are mapped to the blocks in `mod.yaml` - I didn't add any overrides because the defaults here are (AFAIK) suitable for all the default mods.

pchote commented:
> Sometimes when spectating in TS at the closest setting I get some weird artifacts: Black jagged blocks at the bottom of the screen. It seems to be map dependent; I get it consistently on Tactical.

Can you please post a screenshot? I suspect this will be a regression from https://github.com/OpenRA/OpenRA/pull/17308/files#diff-000122b57c1834317a240a39ed02beabR256 in #17308, but would be good to know for sure.

> I think the scroll zoom could benefit from having an acceleration curve.

Sounds reasonable, but I don't have the interest to experiment with values for this. If someone can provide calibration values i'll add them, otherwise we should defer this to a future PR (or if its a blocker I'll drop the smooth zooming from this PR - its a nice extra, but not my focus here).

> At the closest setting there is a noticeable wobble when zooming.

This is caused by the pixel rounding in the viewport scrolling that requires a deeper rework of the Viewport code to fix.  As above, my solution will be to drop the scroll zooming if this is considered a blocker.

tovl commented:
![Screenshot from 2019-12-05 22-08-27](https://user-images.githubusercontent.com/12049990/70274318-363cf600-17ac-11ea-88ac-679effb76dd9.png)

Disregard the colored strip in the middle; that is just a glitch of my OS when making screenshots. It is the black triangles at the bottom.

pchote commented:
Can you confirm that reverting the `/ 2` in the `GetScissorBounds` change I linked above fixes the issue?

tovl commented:
Reverting that doesn't fix it. 

> I can't reproduce the issue locally.

It seems not only map, but also location dependent. It happens for me when I have the camera around the middle spawn position on Tactical.

pchote commented:
Thanks, I can reproduce it now. The problem happens when zooming in by more than a specific threshold on terrain that has a height offset of more than 10 steps. I haven't yet found whatever part of the code this breaks.

pchote commented:
There are two issues here:
* Zooming to large values breaks [this](https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Game/Renderer.cs#L115-L121) assumption.
* Zooming to non-power-of-twos breaks the assumption that `worldBufferSize` and `surfaceBufferSize` have the same aspect ratio relative to their backing sheet.

These may want to be fixed in a separate PR (once we work out *how*), but this one does provide a good testcase.

pchote commented:
Updated:
* Fixed the jagged-bottom-edge issue by rebasing on the updated #17308 
* Changed `unlockMinimumZoomScale` definition
* Added zoom hotkeys: <kbd>[</kbd> and <kbd>]</kbd> because <kbd>-</kbd> is already used and IMO it is awkward for zooming in to require a modifier but zooming out not (at least on my keyboard)

tovl commented:
I can confirm the jagged bottom edge is fixed.

Regarding the wobble: The same wobble is present in the RA shell-map as the camera moves around. It is especially noticeable when setting the camera to "close". Strangely, there is no wobble present when you manually scroll the camera around the map, not even when you do it very slowly using the "joystick" configuration.

> it is awkward for zooming in to require a modifier but zooming out not (at least on my keyboard)

Yeah, realized that after commenting. I guess I was thinking of the numpad + and - keys, but many laptops don't have a numpad so that is not a helpful default.

Regarding the acceleration curve: I've found this article, which provides a reasonable looking curve: 
https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/Scrolling_CHI02.pdf
However, it seems we need to make some slight changes to how we register scroll wheel movement in order to be able to apply a curve. We will need to know the time between subsequent notches of the wheel, which we currently don't track.

That said, the lack of acceleration may not be the main reason why scrolling feels a bit off. I noticed that scroll zooming gets slower the more zoomed in you are. This is not even relative to the base zoom level. For example, with the scroll speed slider somewhere in the middle, it took me 16 notches to get from 1x to 2x on 'furthest'. Yet on 'close' it took me 49 notches to go from 1x to 2x!

pchote commented:
> For example, with the scroll speed slider somewhere in the middle, it took me 16 notches to get from 1x to 2x on 'furthest'. Yet on 'close' it took me 49 notches to go from 1x to 2x!

This is because scrolling adds/subtracts absolute deltas from the zoom: at furthest/native zoom the range between min and max zoom will be 1 (1-2). At the closest zoom we may end up scrolling between 3 and 6, which requires 3 times as many steps.

tovl commented:
That still wouldn't be quite right. That makes it consistent between the different base zoom settings, but the spacing between the notches is still off. This is especially notable in the map editor where the zoom range is larger. The spacing between the hotkey steps is similarly off.

Fortunately, this can be easily remedied. I locally replaced the linear scale with an exponential scale and that completely solved it.

i.e. inside AdjustZoom replace
```
Zoom = (zoom + dz).Clamp(unlockMinZoom ? unlockedMinZoom : minZoom, maxZoom);
```

with

```
Zoom = (zoom * (float)Math.Exp(dz)).Clamp(unlockMinZoom ? unlockedMinZoom : minZoom, maxZoom);
```

This also makes scaling by the base zoom amount redundant.

pchote commented:
Updated:
* Fixed zoom scaling
* Fixed map editor glitches by rebasing on #17431.

tovl commented:
For the record: With the new zoom scaling I don't consider scroll wheel acceleration within the scope of this PR anymore. The wobble in the zoom is also a lot less noticeable with the new scaling, so I don't consider that a blocker. The wobble in the RA shellmap seems to be present on bleed as well, it is just more noticeable with a close camera. Therefore, that is probably out-of-scope as well.

As for the scaling filter itself: I don't have much to say here because it seems to work perfectly. I didn't notice any visual artifacts on any arbitrary zoom level > 1. With zoom < 1 there seems to be some aliasing when moving the map, but that only becomes bothersome with zoom < 0.5 (so only in editor mode; not in spectator mode).

In fact, I couldn't tell any difference in visual fidelity between the optimized default zoom levels and the arbitrary intermediate scroll zoom levels. So I'm beginning to wonder if we really need to have those approximate window heights and wouldn't be better off just using fixed window heights as presets.

Just one more nit: "reset zoom" should be listed under viewport commands, not game commands.

pchote committed:
Add zoom hotkeys.

pchote commented:
Updated:
 * Fixed Undo/Redo button positions in the map editor
 * Moved Reset Zoom hotkey entry into Viewport and grouped with the other two.

pchote commented:
Updated again:
* Replaced/restored the "Also Double Cursor" setting with "Increase Command Cursor Size" which pixel doubles all but the default cursor in both hardware and software cursor modes.

This also accidentally fixes a bug that made hardware cursors always fail on systems with display scales >150%.

pchote commented:
I suspect that there may be an issue hiding in the scaling shader - large zooms are noticeably blurry on my standard definition display, but not on my HiDPI ones.

tovl commented:
Must be a Mac thing then. I have no blurryness at all on my regular 1440p screen in Linux without any scaling by the OS.

tovl commented:
Can you provide screenshots? I do notice some blur at the edges of the pixels, but that is exactly what the shader is supposed to do. If you look at the middle picture in that article, the blur is definitely still noticeable, despite what the article claims. That middle picture corresponds quite well with what I can see on my 1440p monitor.

pchote commented:
It is going to be related to https://github.com/OpenRA/OpenRA/pull/17424/files#diff-3306a4a88c87de34cbb0c547d90a7ab4R166

Launching on HiDPI display (5k / `WindowScale`: 2):
<img width="750" alt="Screenshot 2019-12-08 at 11 36 05" src="https://user-images.githubusercontent.com/167819/70388875-b0899800-19af-11ea-9a3e-c465b532cbe2.png">

Launching on standard display (1440p / `WindowScale`: 1):
![Screenshot 2019-12-08 at 11 36 33](https://user-images.githubusercontent.com/167819/70388888-dd3daf80-19af-11ea-949a-03bc53cf2e46.png)

tovl commented on glsl/combined.frag:
```diff
@@ -64,7 +84,20 @@ vec4 Sample(float samplerIndex, vec2 pos)
+
+		// Offset the sampling point to simulate bilinear intepolation in window instead of texture coordinates
+		// See https://colececil.io/blog/2017/scaling-pixel-art-without-destroying-it/
+		vec2 interp = clamp(offset * 2 * WindowScale, 0.0, .5) + clamp((offset - 1.0) * 2 * WindowScale + .5, 0.0, .5);
```
This is where the problem is: For `WindowScale = 1`, the margin over which we apply the bilinear filter extends to a quarter of the texel width on each side, which results in a blurry image. For `WindowScale = 2`, the margin is only 1/8 of the texel width, so the image is crisper at close zoom levels, but this also makes the aliasing at very far zoom levels worse. 

The margin should instead depend on the actual zoom level and not on `WindowScale` . Replacing with this:
```
vec2 textureSize = Size(vTexSampler.s);
vec2 vUv = coords.st * textureSize;
vec2 alpha = 0.7 * vec2(dFdx(vUv.x), dFdy(vUv.y));		
vec2 offset = fract(vUv);

vec2 interp = clamp(0.5 / alpha * offset, 0.0, .5) + clamp(0.5 / alpha * (offset - 1.0) + .5, 0.0, .5);
coords = (floor(coords.st * textureSize) + interp) / textureSize;
```

solves the problem. See https://csantosbh.wordpress.com/2014/02/05/automatically-detecting-the-texture-filter-threshold-for-pixelated-magnifications/ and https://csantosbh.wordpress.com/2014/01/25/manual-texture-filtering-for-pixelated-games-in-webgl/ (which is where the article in the OP got it's information from in the first place.)

pchote replied:
Aha, good catch. Add another mark to the "incorrect window size vs viewport size assumption" tally :cry:

We can't assume `GL_OES_standard_derivatives` exists in our GLES 3 support, so I instead pass the pixel-texel ratio in to the shader (replacing the old `EnablePixelScaling` bool).

pchote commented:
Updated:
* Fixed scaling filter
* Rebased on top of #17435, which is now a dependency.

pchote committed:
Draw GPS dots using the UI renderers.

pchote commented:
Added a commit to fix the GPS dot rendering glitch discovered by @netnazgul.

netnazgul commented:
Some additional notes on zooming, possibly out of scope for this.
1. Quite accustomed to have "zoom-to-cursor" feature where viewport also centers to mouse pointer during zoom rather than the screen center.
2. Without inertial zoom 90% of values for "zoom speed" are mostly unusable because you need to mash the scroll quite a number of times to get from 0.5x to 2x. I think even on max setting you can't scroll in/out in a single wheel movement.

teinarss commented:
Needs a rebase!

pchote committed:
Implement new viewport size/zoom UI.

pchote commented:
Updated:
- Rebased
- Restored mouse wheel zoom modifier and removed the Scroll Zooming checkbox instead
- Adjusted settings menu layout

pchote commented:
For completeness: https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d5a67c171856ed122f1bda204bf696959b94f73e (@tovl updated the log but didn't comment here).

GraionDilach commented:
This introduced the following crash when I force GL ES on my GPU (GeForce GTX 750 on Win10, driver version 432.00):

```
GL Info Log:
0(88) : error C7011: implicit cast from "int" to "float"
0(90) : error C7011: implicit cast from "ivec2" to "vec2"

System.InvalidProgramException: Compile error in shader object 'D:\OpenRA-AS-dev\glsl\combined.frag'
   at OpenRA.Platforms.Default.Shader.CompileShaderObject(Int32 type, String name)
   at OpenRA.Platforms.Default.Shader..ctor(String name)
   at OpenRA.Platforms.Default.Sdl2GraphicsContext.CreateShader(String name)
   at OpenRA.Renderer..ctor(IPlatform platform, GraphicSettings graphicSettings)
   at OpenRA.Game.Initialize(Arguments args)
GL Info Log:
0(88) : error C7011: implicit cast from "int" to "float"
0(90) : error C7011: implicit cast from "ivec2" to "vec2"

System.InvalidProgramException: Compile error in shader object 'D:\OpenRA-AS-dev\glsl\combined.frag'
   at OpenRA.Platforms.Default.Shader.CompileShaderObject(Int32 type, String name)
   at OpenRA.Platforms.Default.Shader..ctor(String name)
   at OpenRA.Platforms.Default.Sdl2GraphicsContext.CreateShader(String name)
   at OpenRA.Renderer..ctor(IPlatform platform, GraphicSettings graphicSettings)
   at OpenRA.Game.Initialize(Arguments args)
```

pchote commented:
Is this sufficient to fix it?
```diff
diff --git a/glsl/combined.frag b/glsl/combined.frag
index 773386be93..0288f763bc 100644
--- a/glsl/combined.frag
+++ b/glsl/combined.frag
@@ -85,9 +85,9 @@ void main()
 {
        vec2 coords = vTexCoord.st;
 
-       if (AntialiasPixelsPerTexel > 0)
+       if (AntialiasPixelsPerTexel > 0.0)
        {
-               vec2 textureSize = Size(vTexSampler.s);
+               ivec2 textureSize = Size(vTexSampler.s);
                vec2 offset = fract(coords.st * textureSize);
 
                // Offset the sampling point to simulate bilinear intepolation in window coordinates instead of texture coordinates
```

GraionDilach commented:
Sadly not. That then breaks L91 and L99 with the same error message. Changing L91 just further generates errors.

I'll disable GL ES for now instead - I've had it enabled for testing #17087.

pchote commented:
Can you please test with

```diff
diff --git a/glsl/combined.frag b/glsl/combined.frag
index 773386be93..2b63063c91 100644
--- a/glsl/combined.frag
+++ b/glsl/combined.frag
@@ -85,9 +85,9 @@ void main()
 {
        vec2 coords = vTexCoord.st;
 
-       if (AntialiasPixelsPerTexel > 0)
+       if (AntialiasPixelsPerTexel > 0.0)
        {
-               vec2 textureSize = Size(vTexSampler.s);
+               vec2 textureSize = vec2(Size(vTexSampler.s));
                vec2 offset = fract(coords.st * textureSize);
 
                // Offset the sampling point to simulate bilinear intepolation in window coordinates instead of texture coordinates
```
?

GraionDilach commented:
Bingo! That one loads.

Mailaender commented on OpenRA.Game/WorldViewportSizes.cs:
```diff
@@ -0,0 +1,33 @@
+		public readonly int2 FarWindowHeights = new int2(900, 1300);
+
+		public readonly float MaxZoomScale = 2.0f;
+		public readonly int MaxZoomWindowHeight = 240;
```
It would be great to have `[Desc()` documentation on these as they are never used and not quite self explanatory.

QmwJlHuSg9pa commented:
I know I'm rather late to the discussion, though this post here describes a pseudo-bandlimited, cosine-based algorithm that may be an improvement over the current linear-clamp implementation? [Pseudo-bandlimited pixel art filtering in 3D – a mathematical derivation – Maister's Graphics Adventures](https://themaister.net/blog/2018/08/25/pseudo-bandlimited-pixel-art-filtering-in-3d-a-mathematical-derivation/)
And a follow-up post: [Why coverage based pixel art filtering is terrible – Maister's Graphics Adventures](https://themaister.net/blog/2018/08/29/why-coverage-based-pixel-art-filtering-is-terrible/)

On a related note, is pixel aspect ratio correction on the cards? Obviously, this would be more work than simply stretching the play area in one direction... I could open an issue?
